torch/csrc/autograd/saved_variable.cpp

```diff
@@ -57,89 +60,126 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
     // From now on, we can assume the variable is not a leaf and is an output.
 
     is_inplace_on_view_ = is_inplace_on_view;
-    output_nr_ = variable.output_nr();
-    version_counter_ = version_counter;
+
+    if(is_inplace_on_view) {
+      weak_grad_fn_ = variable.grad_fn();
+    }
+
+    save_common_metadata(variable);
 
     // These copies are all shared_ptr copies, so slightly more expensive.
     // Do them here instead of in the init list in case data is undefined.
     data_ = variable.tensor_data();
+  }
+}
 
-    if(is_inplace_on_view) {
-      weak_grad_fn_ = variable.grad_fn();
-    }
+void SavedVariable::save_common_metadata(const Variable& data) {
+  // Saved output number, version counter and fw_grad if needed
 
-    // TODO(albanD) This needs to be updated when moving to multiple levels
-    const auto& fw_grad = variable._fw_grad(/* level */ 0);
-    if (fw_grad.defined()) {
-      fw_grad_ = std::make_shared<ForwardGrad>();
-      fw_grad_->set_value(fw_grad, /* level */ 0);
-    }
+  output_nr_ = data.output_nr();
+  version_counter_ = impl::version_counter(data);
+
+  // TODO(albanD) This needs to be updated when moving to multiple levels
+  const auto& fw_grad = data._fw_grad(/* level */ 0);
+  if (fw_grad.defined()) {
+    fw_grad_ = std::make_shared<ForwardGrad>();
+    fw_grad_->set_value(fw_grad, /* level */ 0);
   }
 }
 
+
+void SavedVariable::reset_data() {
+  hooks_.reset();
+  grad_fn_.reset();
+  data_.reset();
+}
+
 SavedVariable::SavedVariable(const c10::optional<Variable>& variable, bool is_output, bool is_inplace_on_view)
   : SavedVariable(variable.has_value() ? *variable : Variable(), is_output, is_inplace_on_view) {}
 
 Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
-  if (!data_.defined()) {
-    TORCH_CHECK(was_default_constructed_, ERR_BACKWARD_TWICE);
+  if (was_default_constructed_) {
     return Variable();
   }
 
+  if (!data_.defined()) {
+    TORCH_CHECK(hooks_, ERR_BACKWARD_TWICE);
+  }
+
   // We want grad_fn here to provide the most helpful debug message to the user
   // if versions don't match
-  auto grad_fn = saved_original_ ? data_.grad_fn()
-                                 : is_inplace_on_view_ ? weak_grad_fn_.lock()
-                                                       : nullptr;
 
-  if (!saved_original_ && !grad_fn) {
-    TORCH_CHECK(saved_for,"No grad_fn for non-leaf saved variable");
+  auto grad_fn = is_inplace_on_view_ ? weak_grad_fn_.lock()
+                                     : !hooks_ ? saved_original_ ? data_.grad_fn() : nullptr
+                                               : grad_fn_;
+
+  if (!is_leaf_ && !grad_fn) {
+    TORCH_CHECK(saved_for, "No grad_fn for non-leaf saved variable");
```

Comment:
Same for this, is this reachable by the user?

Reply:
This code is not (should not be) reachable from the user.  Will change to TORCH_INTERNAL_CHECK.